src/Stratis.Bitcoin/P2P/PeerAddress.cs

```diff
@@ -0,0 +1,277 @@
+
+                var networkAddress = new NetworkAddress(new IPEndPoint(IPAddress.Parse(this.address), this.addressPort));
+                if (this.addressTime != null)
+                    networkAddress.Time = this.addressTime.Value;
+
+                return networkAddress;
+            }
+        }
+
+        [JsonProperty]
+        private string loopbackAddress;
+        [JsonIgnore]
+        public IPAddress Source
+        {
+            get
+            {
+                if (string.IsNullOrEmpty(this.loopbackAddress))
+                    return null;
+                return IPAddress.Parse(this.loopbackAddress);
+            }
+        }
+
+        #endregion
+
+        #region Connection Data
+
+        [JsonProperty]
+        private int connectionAttempts;
+        [JsonIgnore]
+        public int ConnectionAttempts
+        {
+            get { return this.connectionAttempts; }
+        }
+
+        [JsonProperty]
+        private DateTimeOffset? lastConnectionHandshake;
+        [JsonIgnore]
+        public DateTimeOffset? LastConnectionHandshake
+        {
+            get { return this.lastConnectionHandshake; }
+        }
+
+        [JsonIgnore]
+        public bool IsNew
+        {
+            get
+            {
+                return
+                    this.lastConnectionAttempt == null &&
+                    this.lastConnectionSuccess == null &&
+                    this.lastConnectionHandshake == null;
+            }
+        }
+
+        [JsonProperty]
+        private DateTimeOffset? lastConnectionAttempt;
+        [JsonIgnore]
+        public DateTimeOffset? LastConnectionAttempt
+        {
+            get { return this.lastConnectionAttempt; }
+        }
+
+        [JsonProperty]
+        private DateTimeOffset? lastConnectionSuccess;
+        [JsonIgnore]
+        public DateTimeOffset? LastConnectionSuccess
+        {
+            get { return this.lastConnectionSuccess; }
+        }
+
+        #endregion
+
+        #region Connection Methods
+
+        internal void Attempted(DateTimeOffset peerAttemptedAt)
+        {
+            this.connectionAttempts += 1;
+            this.lastConnectionAttempt = peerAttemptedAt;
+        }
+
+        /// <summary>
+        /// [NBitcoin] Do we need to throttle the update of lastSuccessfulConnect?
+        /// https://github.com/stratisproject/NStratis/blob/2b0fbc3f6b809d92aaf43a8ee12f8baa724e5ccf/NBitcoin/Protocol/AddressManager.cs#L1014
+        /// </summary>
+        internal void Connected(DateTimeOffset peerConnectedAt)
+        {
+            this.addressTime = peerConnectedAt;
+            this.NetworkAddress.Time = peerConnectedAt;
+
+            this.lastConnectionAttempt = null;
+            this.connectionAttempts = 0;
+
+            this.lastConnectionSuccess = peerConnectedAt;
+        }
+
+        /// <summary>
+        /// [NBitcoin] This replaces Good() method
+        /// </summary>
+        internal void Handshaked(DateTimeOffset peerHandshakedAt)
```

Comment:
I feel renaming this to `SetHandshaked` is more readable. 
(same goes for `Connected` to `SetConnected` etc...)

Reply:
Have renamed 👍